User Plan Snapshots

Mythicaeda opened the pull request:
* **Tickets addressed:** Closes #884
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description
<!-- What approach was taken to satisfy the ticket being addressed? What should reviewers be aware of? -->

- The ability to create plan snapshots has been exposed to the user via a Hasura function.
- Another hasura function was made to allow for restoring a plan to a snapshot

## Verification
<!-- How were the changes validated? Were any automated tests added, updated, removed, or re-baselined? -->
- `PermissionsTests` were updated to include the two new function permission keys
- `RestoringSnapshot` DBTests were added to `PlanCollaborationTests`

## Documentation
<!-- What documentation was invalidated by these changes? Which artifacts should reviewers check for accuracy and completeness? -->

[A docs ticket has been made for the API docs changes needs](https://github.com/NASA-AMMOS/aerie-docs/issues/72)
The new keys need to be included in #993 and in the [User-Facing Permissions docs](https://github.com/NASA-AMMOS/aerie-docs/issues/61)

## Future work
<!-- What next steps can we anticipate from here, if any? -->

Mythicaeda committed:
Alter Plan Snapshot Table

- Update Hasura Metadata

Mythicaeda committed:
Add override for create_snapshot

Mythicaeda committed:
Add DBTests

skovati commented on merlin-server/sql/merlin/functions/public/restore_from_snapshot.sql:
```diff
@@ -0,0 +1,102 @@
+        raise exception 'Cannot Restore: Snapshot % is not a snapshot of Plan ''%'' (ID %)',
+				  _snapshot_id, _plan_name, _plan_id;
+			end if;
+		end if;
```
Weird formatting on this `end if` makes it seem like this block is scoped differently. Alternatively, we could maybe unnest these `if`s and just have a flat list of checks / exception raises.

Mythicaeda replied:
Oh, yes, there's something weird where GitHub is drawing the formatting wrong. If you open the file locally, you shouldn't see this. I will update the formatting here and bring it all onto space formatting 👍

skovati replied:
classic tabs vs spaces

Mythicaeda replied:
Formatting has been fixed

skovati commented on merlin-server/sql/merlin/tables/plan_snapshot.sql:
```diff
@@ -8,12 +8,26 @@ create table plan_snapshot(
     on delete set null,
-
   revision integer not null,
-  name text not null,
```
Are these breaking changes?

Mythicaeda replied:
No, because only the DB was able to see or interact with this table, and the changes have been propagated across the DB

skovati replied:
It looks like these were publicly exposed over GQL for regular users in https://github.com/NASA-AMMOS/aerie/pull/1029, so a name change might still count as breaking, not sure 🤔

Mythicaeda replied:
It's not a name change, though. The `name` column (which stored the plan name) was removed alongside a few other columns and the `snapshot_name` column was added.

skovati replied:
Either way, I would imagine removing columns that users might have been querying would still count as a breaking change.

Mythicaeda replied:
Regardless, I know for certain that nobody is currently using this table outside the DB, regardless of its exposure in #1029 ([if you trace it's history you'll see it was accidentally tracked way back when Plan Merge was initially implemented](https://github.com/NASA-AMMOS/aerie/pull/399/files#diff-179570744d2b8431068cbac1cc2a645e1ce9d45226b88d720af05b471bf2896f)). If you feel strongly on this change being breaking, however, feel free to add the `breaking change` label.

skovati replied:
Philosophically it might not feel like a breaking change, but technically I would argue that it meets the definition and should be recorded as accordingly in our e.g. upgrade docs, just in case.

Kinda the downside of exposing our entire DB schema as an API 🤷

skovati replied:
We can check with @camargo when we do the release next week if this should be included in the upgrade docs.

skovati commented on merlin-server/sql/merlin/tables/plan_snapshot.sql:
```diff
@@ -8,12 +8,26 @@ create table plan_snapshot(
-  duration interval not null,
-  start_time timestamptz not null
+
+  snapshot_name text null,
+  taken_by text null,
```
```suggestion
  snapshot_name text,
  taken_by text,
```

According to [postgres docs](https://www.postgresql.org/docs/current/sql-createtable.html):
"This [null] clause is only provided for compatibility with non-standard SQL databases. Its use is discouraged in new applications."

Mythicaeda replied:
`null` has been removed from the new columns 👍

skovati commented on deployment/hasura/metadata/databases/AerieMerlin/tables/public_plan_snapshot_activities.yaml:
```diff
@@ -16,3 +16,26 @@ array_relationships:
+  - role: aerie_admin
+    permission:
+      columns: [snapshot_id, id, name, source_scheduling_goal_id, created_at, last_modified_at, start_offset, type,
+                arguments, last_modified_arguments_at, metadata, anchor_id, anchored_to_start]
```
Make sure to add `last_modified_by` here once https://github.com/NASA-AMMOS/aerie/pull/1041 is merged 👍

Mythicaeda replied:
Done!

skovati reviewed (commented):
Looks really good to me so far! I'll do a final approval review tomorrow once this is rebased on top of https://github.com/NASA-AMMOS/aerie/pull/1041.

Mythicaeda committed:
Add Restore Snapshot Functionality

Mythicaeda committed:
Add Hasura function for taking/restoring snapshots

Mythicaeda committed:
Add `create_snapshot` and `restore_snapshot` to `function_permission_key`

- Add new keys to `user_role_permission`
- Alphabetizes the JSONs

Mythicaeda committed:
Update PermissionsTest

- Add new keys/cases
- Alphabetize `getFunctionReturnsAssignedValueUser`
- Correct `testMissionModel` to `testMissionModelOwner`

skovati commented on merlin-server/sql/merlin/functions/public/restore_from_snapshot.sql:
```diff
@@ -0,0 +1,105 @@
+		       psa.start_offset, psa.type, psa.arguments, psa.last_modified_arguments_at, psa.metadata,
+		       psa.anchor_id, psa.anchored_to_start
+		from plan_snapshot_activities psa
+		where psa.snapshot_id = _snapshot_id
+		on conflict (id, plan_id) do update
+		-- 'last_modified's are skipped during update, as triggers will overwrite them to now()
+		set name = excluded.name,
+		    source_scheduling_goal_id = excluded.source_scheduling_goal_id,
+		    created_at = excluded.created_at,
+		    last_modified_by = excluded.last_modified_by,
+		    start_offset = excluded.start_offset,
+		    type = excluded.type,
+		    arguments = excluded.arguments,
+		    metadata = excluded.metadata,
+		    anchor_id = excluded.anchor_id,
+		    anchored_to_start = excluded.anchored_to_start;
+
+		-- Tags
+		delete from metadata.activity_directive_tags adt
+		using diff d
+		where (adt.directive_id, adt.plan_id) = (d.activity_id, _plan_id);
+
+		insert into metadata.activity_directive_tags(directive_id, plan_id, tag_id)
+		select sat.directive_id, _plan_id, sat.tag_id
+		from metadata.snapshot_activity_tags sat
+		where sat.snapshot_id = _snapshot_id
+		on conflict (directive_id, plan_id, tag_id) do nothing;
+
+		-- Presets
+		delete from preset_to_directive
+		  where plan_id = _plan_id;
+		insert into preset_to_directive(preset_id, activity_id, plan_id)
+			select pts.preset_id, pts.activity_id, _plan_id
+			from preset_to_snapshot_directive pts
+			where pts.snapshot_id = _snapshot_id
+			on conflict (activity_id, plan_id)
+			do update	set preset_id = excluded.preset_id;
+
+		-- Clean up
+		drop table diff;
```
Nonblocking: there's still some weird formatting for me here, is github displaying 4 space tabs while your editor displays 2 space tabs or something?

Mythicaeda replied:
GH is displaying 4 spaces tabs while the editor displays two spaced tabs. I'm not sure how to fix this, as this a really recent bug in either IntelliJ or GH (I'm leaning towards GH, as I checked my IntelliJ settings and confirmed its using a 2-space tab)

skovati replied:
¯\\_(ツ)_/¯

skovati commented on deployment/hasura/metadata/databases/AerieMerlin/tables/public_plan_snapshot_activities.yaml:
```diff
@@ -16,3 +16,26 @@ array_relationships:
+select_permissions:
+  - role: aerie_admin
+    permission:
+      columns: [snapshot_id, id, name, source_scheduling_goal_id, created_at, last_modified_at, start_offset, type,
```
Looks like this is still missing `last_modified_by`

Mythicaeda replied:
Huh, that should've been brought in in your PR. I'll update it 👍

skovati replied:
I guess this whole section of permissions didn't exist before, so it was easy to miss

Mythicaeda replied:
Ah, that explains it. Fix has been made.

skovati commented on merlin-server/sql/merlin/functions/hasura/snapshot_functions.sql:
```diff
@@ -0,0 +1,44 @@
+create table hasura_functions.create_snapshot_return_value(snapshot_id integer);
```
Are these intermediate tables necessary for hasura functions? I.e we can't just return `_snapshot_id` as an integer?

Mythicaeda replied:
Yes, very sadly. [See the return type of `Supported SQL Functions` in the docs.](https://hasura.io/docs/latest/schema/postgres/custom-functions/)

skovati commented on merlin-server/sql/merlin/functions/hasura/snapshot_functions.sql:
```diff
@@ -0,0 +1,44 @@
+    raise exception 'Snapshot name cannot be null.';
+  end if;
+
+  select create_snapshot(_plan_id, _snapshot_name, _snapshotter) into _snapshot_id;
+  return row(_snapshot_id)::hasura_functions.create_snapshot_return_value;
```
(continuing my above comment) so that this could instead be something like

```suggestion
  return query create_snapshot(_plan_id, _snapshot_name, _snapshotter)
```

Mythicaeda replied:
Forget `return query`, if we could just return non-tables, we could change that into `return create_snapshot`

skovati commented on merlin-server/sql/merlin/functions/public/restore_from_snapshot.sql:
```diff
@@ -0,0 +1,105 @@
+		from plan_snapshot_activities psa
+		where psa.snapshot_id = _snapshot_id
+		on conflict (id, plan_id) do update
+		-- 'last_modified's are skipped during update, as triggers will overwrite them to now()
```
This comment is a bit confusing, since we do reference a `last_modified_by` here.

Mythicaeda replied:
Comment has been tweaked

skovati reviewed (approved):
Looks great! Just some nonblocking comments, none of them would change functionality.

Mythicaeda committed:
Update Hasura Functions Metadata

- alphabetize functions.yaml

skovati commented:
🚀
